src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+        internal class PendingBranch
         {
-            if (_trackUnassignments)
+            public readonly BoundNode Branch;
+            public TLocalState State;
+            public LabelSymbol Label
             {
-                Optional<TLocalState> oldTryState = _tryState;
-                _tryState = AllBitsSet();
-                VisitTryBlock(tryBlock, node, ref tryState);
-                var tempTryStateValue = _tryState.Value;
-                IntersectWith(ref tryState, ref tempTryStateValue);
-                if (oldTryState.HasValue)
+                get
                 {
-                    var oldTryStateValue = oldTryState.Value;
-                    IntersectWith(ref oldTryStateValue, ref tempTryStateValue);
-                    oldTryState = oldTryStateValue;
+                    if (Branch == null) return null;
+                    switch (Branch.Kind)
+                    {
+                        case BoundKind.GotoStatement: return ((BoundGotoStatement)Branch).Label;
+                        case BoundKind.ConditionalGoto: return ((BoundConditionalGoto)Branch).Label;
+                        case BoundKind.BreakStatement: return ((BoundBreakStatement)Branch).Label;
+                        case BoundKind.ContinueStatement: return ((BoundContinueStatement)Branch).Label;
+                        case BoundKind.PatternSwitchLabel: return ((BoundPatternSwitchLabel)Branch).Label;
+                        default: return null;
+                    }
                 }
+            }
 
-                _tryState = oldTryState;
+            public PendingBranch(BoundNode branch, TLocalState state)
+            {
+                this.Branch = branch;
+                this.State = state.Clone();
             }
-            else
+        }
+
+        /// <summary>
+        /// Perform a single pass of flow analysis.  Note that after this pass,
+        /// this.backwardBranchChanged indicates if a further pass is required.
+        /// </summary>
+        protected virtual ImmutableArray<PendingBranch> Scan(ref bool badRegion)
+        {
+            var oldPending = SavePending();
+            Visit(methodMainNode);
+            this.Unsplit();
+            RestorePending(oldPending);
+            if (_trackRegions && regionPlace != RegionPlace.After) badRegion = true;
+            ImmutableArray<PendingBranch> result = RemoveReturns();
+            return result;
+        }
+
+        protected ImmutableArray<PendingBranch> Analyze(ref bool badRegion)
+        {
+            ImmutableArray<PendingBranch> returns;
+            do
             {
-                VisitTryBlock(tryBlock, node, ref tryState);
+                // the entry point of a method is assumed reachable
+                regionPlace = RegionPlace.Before;
+                this.State = TopState();
+                PendingBranches.Clear();
+                if (_trackExceptions) PendingBranches.Add(new PendingBranch(null, TopState()));
```

Comment:
fwiw, i find the single-line ifs jarring each time i run into them.  i actually read them every time as if the statement below them is conditioned on them.

Reply:
this check could use documentation.

Reply:
Yes, the coding guidelines don't require curly braces, but do require the controlled statement be on its own line.  This code is coped from `PreciseAbstractFlowAnalysis`, was written before we had coding guidelines, and hasn't been changed in quite a few years.

---
In reply to: [236567282](https://github.com/dotnet/roslyn/pull/31389#discussion_r236567282) [](ancestors = 236567282)